### DISCUSSION TITLE
v3 to v4 conversion questions

### DISCUSSION THREAD

**Discussion by zieglerjm** [wixtoolset/issues#7850] [Questions]
I am trying to convert a WIX v3 project to v4.  How do I convert "$(var.XXX)"?

<Package Name="!(loc.ManufacturerName) **$(var.ProductName)** !(loc.ClientSP) **$(var.BuildVersion)**" Language="!(loc.Language)" Version="**$(var.BuildVersion)**" Manufacturer="!(loc.ManufacturerFullName)"

![image](https://github.com/wixtoolset/issues/assets/25181017/66f3b504-832f-49ac-8495-108e75cd7ec9)

The v3 project was using Target BeforeBuild -> PropertyGroup -> DefineConstants.  So far, I have not had an error on that in the v4 project, so I am assuming it is working?

```
<Project Sdk="WixToolset.Sdk/4.0.2">
    <Target Name="BeforeBuild">
        <PropertyGroup>
            <DefineConstants>$(DefineConstants);BuildVersion=$(TargetVersion);ProductName=$(ProductName)</DefineConstants>
        </PropertyGroup>
    </Target>
</Project>
```

I also tried this with not change

- <DefineConstants>$(DefineConstants);BuildVersion=**1.0.0**;ProductName=**Test**</DefineConstants>

**Comment by zieglerjm** [comment]
After moving the `<DefineConstants>` into the `<PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|x64'">` group, that error went away.  So, now it doesn't like my `<Publish Dialog=.../>`.  Unfortunately, the person that did the WIX 3 work has left, so I am new to WIX try to upgrade his code.

Error: WIX0094 The identifier 'Dialog:InstallDirDlg' could not be found. Ensure you have typed the reference correctly and that all the necessary inputs are provided to the linker.

`<UI>
   ...
   <Publish Dialog="WelcomeDlg" Control="Next" Event="NewDialog" Value="InstallDirDlg" Order="2" />
   <Publish Dialog="InstallDirDlg" Control="Back" Event="NewDialog" Value="WelcomeDlg" Order="2" />
   ...
</UI>`

**Reply by wmanning** [reply]
Seems like the previous person used custom dialogs.  Those have changed from Wix 3 to Wix4 according to below documentation:
-  https://wixtoolset.org/docs/fourthree/faqs/#referencing-the-standard-wixui-dialog-sets
-  https://wixtoolset.org/docs/fourthree/faqs/#converting-custom-wixui-dialog-sets
Did you review that documentation?

I don't use the standard or custom dialogs so not sure if that's the issue or not.

**Comment by zieglerjm** [comment]
After installing random NUGET packages I got a new error - WIX0094	The identifier 'Dialog:WIXUI_EXITDIALOGOPTIONALTEXT' could not be found. Ensure you have typed the reference correctly and that all the necessary inputs are provided to the linker.

`<ItemGroup>
    <PackageReference Include="WixToolset.Bal.wixext" Version="4.0.2" />
    <PackageReference Include="WixToolset.Heat" Version="4.0.2" />
    <PackageReference Include="WixToolset.UI.wixext" Version="4.0.2" />
    <PackageReference Include="WixToolset.Util.wixext" Version="4.0.2" />
  </ItemGroup>`

`<UI>
   <Publish Dialog="ExitDialog" Control="Finish" Event="DoAction" Value="LaunchConfigMgr" Condition="WIXUI_EXITDIALOGOPTIONALCHECKBOX = 1 and NOT Installed" />
   <Publish Dialog="WIXUI_EXITDIALOGOPTIONALTEXT" Control="Finish" Event="DoAction" Value="LaunchConfigMgr" Condition="WIXUI_EXITDIALOGOPTIONALCHECKBOX = 1 and NOT Installed" />
</UI>`

**Reply by wmanning** [reply]
> After installing random NUGET packages

???

**Comment by zieglerjm** [comment]
Currently I am trying to get the ComponentGroup & ComponentRef to work.  It would be nice if there was a working example somewhere I could follow.

**Comment by zieglerjm** [comment]
Looks like there is a difference on how heat is working in v4, or a bug.  In v3 heat was seetting the DirectoryRef to the value in ComponentGroup that is was attached to.  In v4, it is not.  How do I fix this?

Error message: WIX0094	The identifier '**Directory:{Project}.Content**' could not be found
```
<ComponentGroup Id="ABC" Directory="INSTALLDIR">
            <ComponentGroupRef Id="{Project}.Content" />
</ComponentGroup>

_{Project}.wxs
v3: <DirectoryRef Id="INSTALLDIR">
v4: <DirectoryRef Id="{Project}.Content">

```
